spinoso-time/src/time/tzrs/mod.rs

```diff
@@ -0,0 +1,259 @@
+use core::cmp::Ordering;
+use core::hash::{Hash, Hasher};
+
+use tz::datetime::DateTime;
+
+mod build;
+mod convert;
+mod math;
+mod offset;
+mod parts;
+mod timezone;
+mod to_a;
+
+pub use offset::Offset;
+pub use to_a::ToA;
+
+use crate::NANOS_IN_SECOND;
+
+/// A wrapper around [`tz::datetime::DateTime`] which contains everything needed for date creation and
+/// conversion to match the ruby spec. Seconds and Subseconds are stored independently as i64 and
+/// u32 respectively, which gives enough granularity to meet the ruby [`Time`] spec.
+///
+/// [`Time`]: https://ruby-doc.org/core-2.6.3/Time.html
+#[must_use]
+#[derive(Debug, Copy, Clone)]
+pub struct Time {
+    /// A wrapper around [`tz::datetime::DateTime`] to provide date and time formatting
+    inner: DateTime,
+    /// The offset to used for the provided _time_
+    offset: Offset,
+}
+
+impl Hash for Time {
+    #[inline]
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_i128(self.inner.total_nanoseconds());
+    }
+}
+
+impl PartialEq for Time {
+    fn eq(&self, other: &Time) -> bool {
+        self.inner.total_nanoseconds() == other.inner.total_nanoseconds()
+    }
+}
+
+impl Eq for Time {}
+
+impl PartialOrd for Time {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Time {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.inner.total_nanoseconds().cmp(&other.inner.total_nanoseconds())
+    }
+}
+
+// constructors
+impl Time {
+    /// Returns a new Time from the given values in the provided `offset`.
+    ///
+    /// Can be used to implment ruby [`Time#new`] (using a [`Timezone`] Object)
+    ///
+    /// Note: During DST transitions, a specific time can be ambiguous. This method will always pick the earliest date.
+    ///
+    /// # Examples
+    /// ```
+    /// use spinoso_time::tzrs::{Time, Offset};
+    /// use tzdb::time_zone::pacific::AUCKLAND;
+    /// let offset = Offset::tz(AUCKLAND);
+    /// let t = Time::new(2022, 9, 25, 1, 30, 0, 0, offset);
+    /// ```
+    ///
+    /// [`Time#new`]: https://ruby-doc.org/core-2.6.3/Time.html#method-c-new
+    /// [`Timezone`]: https://ruby-doc.org/core-2.6.3/Time.html#class-Time-label-Timezone+argument
+    #[inline]
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        year: i32,
+        month: u8,
+        day: u8,
+        hour: u8,
+        minute: u8,
+        second: u8,
+        nanoseconds: u32,
+        offset: Offset,
+    ) -> Self {
+        let tz = offset.time_zone_ref();
+        let found_date_times = DateTime::find(year, month, day, hour, minute, second, nanoseconds, tz)
+            .expect("Could not find a matching DateTime for this timezone");
```

Comment:
I took a peek at https://github.com/b-n/artichoke/pull/171 and it looks like you have a plan for making these functions fallible and removing the `expect`s. Thank you!

Reply:
:100: correct. writing these `.expect`s hurt a little at the time, but the future glory will prevail!